File: sdk/python/foodblock/forward.py
"""
FoodBlock Forward Traversal — downstream graph navigation.

While chain() follows refs backwards (provenance), forward() follows
refs forward (impact). Essential for recall operations: "which products
used this contaminated ingredient?"

Usage:
    result = await forward(wheat_hash, resolve_forward)
    # result = {'referencing': [{'block': {...}, 'role': 'inputs'}, ...], 'count': 3}

    result = await recall(contaminated_hash, resolve_forward)
    # result = {'affected': [...], 'depth': 4, 'paths': [[hash1, hash2, ...], ...]}
"""

from typing import Callable, Optional


async def forward(hash_val: str, resolve_forward: Callable) -> dict:
    """
    Find all blocks that reference a given hash in any ref field.

    Args:
        hash_val: hash to search for
        resolve_forward: async function(hash) -> list of blocks that reference this hash

    Returns:
        {'referencing': [{'block': block, 'role': str}, ...], 'count': int}
    """
    children = await resolve_forward(hash_val)
    referencing = []

    for block in children:
        refs = block.get("refs", {})
        for role, ref in refs.items():
            hashes = ref if isinstance(ref, list) else [ref]
            if hash_val in hashes:
                referencing.append({"block": block, "role": role})

    return {"referencing": referencing, "count": len(referencing)}


async def recall(
    source_hash: str,
    resolve_forward: Callable,
    max_depth: int = 50,
    types: Optional[list] = None,
    roles: Optional[list] = None,
) -> dict:
    """
    Trace a contamination/recall path downstream via BFS.

    Args:
        source_hash: the contaminated/recalled block hash
        resolve_forward: async function(hash) -> list of blocks referencing this hash
        max_depth: maximum traversal depth (default 50)
        types: optional list of type prefixes to filter (e.g. ['substance'])
        roles: optional list of ref roles to filter

    Returns:
        {'affected': [blocks], 'depth': int, 'paths': [[hash, ...], ...]}
    """
    visited = set()
    visited.add(source_hash)
    affected = []
    paths = []
    max_depth_seen = 0

    # BFS queue entries: (hash_to_expand, current_depth, path_so_far)
    queue = [(source_hash, 0, [source_hash])]

    while queue:
        current_hash, depth, path = queue.pop(0)

        if depth >= max_depth:
            continue

        children = await resolve_forward(current_hash)

        for block in children:
            block_hash = block.get("hash")
            if not block_hash or block_hash in visited:
                continue

            # Check which ref roles connect back to current_hash
            refs = block.get("refs", {})
            matched = False
            for role, ref in refs.items():
                hashes = ref if isinstance(ref, list) else [ref]
                if current_hash in hashes:
                    if roles and role not in roles:
                        continue
                    matched = True
                    break

            if not matched:
                continue

            visited.add(block_hash)

            # Apply type prefix filter
            block_type = block.get("type", "")
            if types:
                if not any(block_type.startswith(t) for t in types):
                    # Still traverse through non-matching types to find
                    # matching blocks deeper in the graph
                    new_path = path + [block_hash]
                    next_depth = depth + 1
                    queue.append((block_hash, next_depth, new_path))
                    continue

            new_path = path + [block_hash]
            next_depth = depth + 1

            affected.append(block)
            paths.append(new_path)
            if next_depth > max_depth_seen:
                max_depth_seen = next_depth

            queue.append((block_hash, next_depth, new_path))

    return {"affected": affected, "depth": max_depth_seen, "paths": paths}


async def downstream(
    ingredient_hash: str,
    resolve_forward: Callable,
    max_depth: int = 50,
) -> list:
    """
    Find all downstream substance blocks that use a given ingredient.
    Convenience wrapper around recall() filtering for substance.* types.

    Args:
        ingredient_hash: hash of the ingredient block
        resolve_forward: async function(hash) -> list of blocks referencing this hash
        max_depth: max depth to search

    Returns:
        list of substance blocks
    """
    result = await recall(
        ingredient_hash,
        resolve_forward,
        max_depth=max_depth,
        types=["substance"],
    )
    return result["affected"]

File: sdk/swift/Sources/Forward.swift
import Foundation

/// FoodBlock Forward Traversal — downstream graph navigation.
///
/// While `FoodBlockChain.chain()` follows refs backwards (provenance),
/// `FoodBlockForward.forward()` follows refs forward (impact). Essential for
/// recall operations: "which products used this contaminated ingredient?"

public enum FoodBlockForward {

    // MARK: - Result Types

    /// A block that references the searched hash, paired with the ref role used.
    public struct ReferencingEntry {
        public let block: FoodBlock
        public let role: String
    }

    /// Result of a forward lookup: all blocks referencing a given hash.
    public struct ForwardResult {
        public let referencing: [ReferencingEntry]
        public let count: Int
    }

    /// Result of a recall traversal: all affected blocks downstream of a source.
    public struct RecallResult {
        public let affected: [FoodBlock]
        public let depth: Int
        public let paths: [[String]]
    }

    // MARK: - Forward

    /// Find all blocks that reference a given hash in any ref field.
    ///
    /// - Parameters:
    ///   - hash: The 64-character hex hash to search for.
    ///   - resolveForward: Async function that returns all blocks referencing a given hash.
    /// - Returns: A `ForwardResult` with referencing blocks and count.
    public static func forward(
        hash: String,
        resolveForward: @escaping (String) async -> [FoodBlock]
    ) async -> ForwardResult {
        let blocks = await resolveForward(hash)

        var referencing: [ReferencingEntry] = []

        for block in blocks {
            for (role, ref) in block.refs {
                let hashes: [String]
                if let s = ref.value as? String {
                    hashes = [s]
                } else if let arr = ref.value as? [Any] {
                    hashes = arr.compactMap { $0 as? String }
                } else {
                    continue
                }

                if hashes.contains(hash) {
                    referencing.append(ReferencingEntry(block: block, role: role))
                }
            }
        }

        return ForwardResult(referencing: referencing, count: referencing.count)
    }

    // MARK: - Recall

    /// Trace a contamination/recall path downstream.
    ///
    /// Starting from a source block (e.g., contaminated ingredient), follow all
    /// forward references recursively using BFS to find every affected block.
    ///
    /// - Parameters:
    ///   - sourceHash: The contaminated/recalled block hash.
    ///   - resolveForward: Async function that returns blocks referencing a given hash.
    ///   - maxDepth: Maximum traversal depth (default 50).
    ///   - types: Optional type filter. Supports wildcards: "substance.*" matches "substance.product".
    ///   - roles: Optional ref role filter. Only follow connections through these roles.
    /// - Returns: A `RecallResult` with affected blocks, max depth reached, and traversal paths.
    public static func recall(
        sourceHash: String,
        resolveForward: @escaping (String) async -> [FoodBlock],
        maxDepth: Int = 50,
        types: [String]? = nil,
        roles: [String]? = nil
    ) async -> RecallResult {
        var visited = Set<String>()
        var affected: [FoodBlock] = []
        var maxDepthReached = 0
        var paths: [[String]] = []

        // BFS queue entries
        struct QueueEntry {
            let hash: String
            let depth: Int
            let path: [String]
        }

        var queue: [QueueEntry] = [QueueEntry(hash: sourceHash, depth: 0, path: [sourceHash])]
        visited.insert(sourceHash)

        while !queue.isEmpty {
            let entry = queue.removeFirst()

            if entry.depth >= maxDepth { continue }

            let blocks = await resolveForward(entry.hash)

            for block in blocks {
                if block.hash.isEmpty { continue }
                if visited.contains(block.hash) { continue }

                // Determine which ref roles connect this block to the current hash
                var matchingRoles: [String] = []
                for (role, ref) in block.refs {
                    let hashes: [String]
                    if let s = ref.value as? String {
                        hashes = [s]
                    } else if let arr = ref.value as? [Any] {
                        hashes = arr.compactMap { $0 as? String }
                    } else {
                        continue
                    }
                    if hashes.contains(entry.hash) {
                        matchingRoles.append(role)
                    }
                }

                // Filter by roles if specified
                if let roles = roles, !matchingRoles.isEmpty {
                    let hasMatchingRole = matchingRoles.contains { r in roles.contains(r) }
                    if !hasMatchingRole { continue }
                }

                // Filter by types if specified (supports wildcards like "substance.*")
                if let types = types {
                    let matchesType = types.contains { t in
                        if t.hasSuffix(".*") {
                            let prefix = String(t.dropLast(1)) // drop the "*", keep the "."
                            return block.type.hasPrefix(prefix)
                        }
                        return block.type == t
                    }
                    if !matchesType { continue }
                }

                visited.insert(block.hash)

                let blockPath = entry.path + [block.hash]
                let currentDepth = entry.depth + 1

                if currentDepth > maxDepthReached {
                    maxDepthReached = currentDepth
                }

                affected.append(block)
                paths.append(blockPath)

                queue.append(QueueEntry(hash: block.hash, depth: currentDepth, path: blockPath))
            }
        }

        return RecallResult(affected: affected, depth: maxDepthReached, paths: paths)
    }

    // MARK: - Downstream

    /// Find all downstream products of a given input.
    ///
    /// Convenience wrapper around `recall()` that filters for substance.* types by default.
    ///
    /// - Parameters:
    ///   - ingredientHash: Hash of the ingredient to trace downstream.
    ///   - resolveForward: Async function that returns blocks referencing a given hash.
    ///   - types: Block types to include. Defaults to `["substance.*"]`.
    /// - Returns: Array of substance blocks that use this ingredient (directly or indirectly).
    public static func downstream(
        ingredientHash: String,
        resolveForward: @escaping (String) async -> [FoodBlock],
        types: [String]? = nil
    ) async -> [FoodBlock] {
        let result = await recall(
            sourceHash: ingredientHash,
            resolveForward: resolveForward,
            types: types ?? ["substance.*"]
        )
        return result.affected
    }
}
